boto/connection.py

```diff
@@ -1062,16 +1063,160 @@ def build_base_http_request(self, method, path, auth_path,
+        new_endpoint = self._get_correct_s3_endpoint_from_response(
+            request,
+            err,
+            get_header
+        )
+        if not new_endpoint:
+            return None
+
+        msg = 'This request was sent to %s, ' % request.host
+        msg += 'when it should have been sent to %s. ' % new_endpoint
+        request.host = new_endpoint
+
+        new_host = self._get_s3_host(new_endpoint)
+        if new_host and new_host != self.host:
+            msg += 'This error may have arisen because your S3 host, ' 
+            msg += 'currently %s, is configured incorrectly. ' % self.host
+            msg += 'Please change your configuration to use %s ' % new_host
+            msg += 'to avoid multiple unnecessary redirects '
+            msg += 'and signing attempts.'
+            self.host = new_host
+
+        boto.log.debug(msg)
+        return request
+
+    def _get_request_for_s3_retry(self, http_request, response, err):
+        if response:
+            body = response.read()
+            if body:
+                body = body.decode('utf-8')
+            # wrap response in an S3 error so its xml body is parsed.
+            err = S3ResponseError(response.status, response.reason, body)
+            return self._change_s3_host_from_error(
+                http_request,
+                err,
+                response.getheader
+            )
+        elif err:
+            return self._change_s3_host_from_error(
+                http_request,
+                err
+            )
+
     def make_request(self, method, path, headers=None, data='', host=None,
                      auth_path=None, sender=None, override_num_retries=None,
                      params=None, retry_handler=None):
-        """Makes a request to the server, with stock multiple-retry logic."""
+        """Make a request to the server.
+        Include logic for retrying on s3 region errors.
+        """
         if params is None:
             params = {}
         http_request = self.build_base_http_request(method, path, auth_path,
                                                     params, headers, data, host)
-        return self._mexe(http_request, sender, override_num_retries,
-                          retry_handler=retry_handler)
+        response, err = None, None    
+        try:
+            response = self._mexe(http_request, sender, override_num_retries,
+                                  retry_handler=retry_handler)
+        except S3ResponseError as e:
+            # Sender functions passed into _mexe often call
+            # connection.getresponse directly, and raise their own exceptions
+            # if a request fails. See _send_file_internal in s3/key.py. The
+            # exception can contain information useful for a retry, however,
+            # so it needs to be in scope for the retry logic below.
+            err = e
+
+        status = (response or err).status
```

Comment:
So is it possible that we might get an error in the response as well as an exception?

Reply:
Yeah, actually the only cases I know of that we have an already packaged exception at this point are
1. If an HTTP exception gets thrown after multiple retries
2. If the sender function (used with copy) throws an error.

I'm pretty sure that in all of the other cases we will just have the error information in the api response at this point.